File: ddd-domain-repo-jpa/src/main/kotlin/com/only4/cap4k/ddd/domain/repo/impl/DefaultAggregateSupervisor.java
package com.only4.cap4k.ddd.domain.repo.impl;

import com.only4.cap4k.ddd.core.domain.aggregate.Aggregate;
import com.only4.cap4k.ddd.core.domain.aggregate.AggregateFactorySupervisor;
import com.only4.cap4k.ddd.core.domain.aggregate.AggregatePayload;
import com.only4.cap4k.ddd.core.domain.aggregate.Id;
import com.only4.cap4k.ddd.core.domain.repo.AggregatePredicate;
import com.only4.cap4k.ddd.core.domain.repo.AggregateSupervisor;
import com.only4.cap4k.ddd.core.domain.repo.RepositorySupervisor;
import com.only4.cap4k.ddd.core.share.OrderInfo;
import com.only4.cap4k.ddd.core.share.PageData;
import com.only4.cap4k.ddd.core.share.PageParam;
import com.only4.cap4k.ddd.core.share.misc.ClassUtils;
import lombok.RequiredArgsConstructor;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * 默认聚合管理器
 *
 * @author binking338
 * @date 2025/1/12
 */
@RequiredArgsConstructor
public class DefaultAggregateSupervisor implements AggregateSupervisor {
    final RepositorySupervisor repositorySupervisor;

    private static <AGGREGATE extends Aggregate<?>> AGGREGATE newInstance(Class<AGGREGATE> clazz, Object entity) {
        try {
            Aggregate aggregate = clazz.getConstructor().newInstance();
            aggregate._wrap(entity);
            return (AGGREGATE)aggregate;
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    @Override
    public <AGGREGATE extends Aggregate<ENTITY>, ENTITY_PAYLOAD extends AggregatePayload<ENTITY>, ENTITY> AGGREGATE create(Class<AGGREGATE> clazz, ENTITY_PAYLOAD payload) {
        ENTITY entity = AggregateFactorySupervisor.getInstance().create(payload);
        return newInstance(clazz, entity);
    }

    @Override
    public <AGGREGATE extends Aggregate<ENTITY>, ENTITY> List<AGGREGATE> getByIds(Iterable<Id<AGGREGATE, ?>> ids, boolean persist) {
        if (!ids.iterator().hasNext()) {
            return Collections.emptyList();
        }
        Class<AGGREGATE> aggregateClass = (Class<AGGREGATE>)ClassUtils.resolveGenericTypeClass(ids.iterator().next(), 0, Id.class, Id.Default.class);
        AggregatePredicate<AGGREGATE, ENTITY> aggregatePredicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicate.byIds(
                aggregateClass,
                StreamSupport.stream(ids.spliterator(), false).map(Id::getValue).collect(Collectors.toList())
        );
        return find(aggregatePredicate, persist);
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> List<AGGREGATE> find(AggregatePredicate<AGGREGATE, ?> aggregatePredicate, Collection<OrderInfo> orders, boolean persist) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        List<?> entities = repositorySupervisor.find(predicate, orders, persist);
        return entities.stream().map(e -> newInstance(clazz, e)).collect(Collectors.toList());
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> List<AGGREGATE> find(AggregatePredicate<AGGREGATE, ?> aggregatePredicate, PageParam pageParam, boolean persist) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        List<?> entities = repositorySupervisor.find(predicate, pageParam, persist);
        return entities.stream().map(e -> newInstance(clazz, e)).collect(Collectors.toList());
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> Optional<AGGREGATE> findOne(AggregatePredicate<AGGREGATE, ?> aggregatePredicate, boolean persist) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        Optional<?> entity = repositorySupervisor.findOne(predicate, persist);
        return entity.map(e -> newInstance(clazz, e));
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> Optional<AGGREGATE> findFirst(AggregatePredicate<AGGREGATE, ?> aggregatePredicate, Collection<OrderInfo> orders, boolean persist) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        Optional<?> entity = repositorySupervisor.findFirst(predicate, orders, persist);
        return entity.map(e -> newInstance(clazz, e));
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> PageData<AGGREGATE> findPage(AggregatePredicate<AGGREGATE, ?> aggregatePredicate, PageParam pageParam, boolean persist) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        PageData<?> entities = repositorySupervisor.findPage(predicate, pageParam, persist);
        return entities.transform(e -> newInstance(clazz, e));
    }

    @Override
    public <AGGREGATE extends Aggregate<ENTITY>, ENTITY> List<AGGREGATE> removeByIds(Iterable<Id<AGGREGATE, ?>> ids) {
        if (!ids.iterator().hasNext()) {
            return Collections.emptyList();
        }
        Class<AGGREGATE> aggregateClass = (Class<AGGREGATE>)ClassUtils.resolveGenericTypeClass(ids.iterator().next(), 0, Id.class, Id.Default.class);
        AggregatePredicate<AGGREGATE, ?> aggregatePredicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicate.byIds(
                aggregateClass,
                StreamSupport.stream(ids.spliterator(), false).map(Id::getValue).collect(Collectors.toList())
        );
        return remove(aggregatePredicate);
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> List<AGGREGATE> remove(AggregatePredicate<AGGREGATE, ?> aggregatePredicate) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        List<?> entities = repositorySupervisor.remove(predicate);
        return entities.stream().map(e -> newInstance(clazz, e)).collect(Collectors.toList());
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> List<AGGREGATE> remove(AggregatePredicate<AGGREGATE, ?> aggregatePredicate, int limit) {
        Class<AGGREGATE> clazz = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.reflectAggregateClass(aggregatePredicate);
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        List<?> entities = repositorySupervisor.remove(predicate, limit);
        return entities.stream().map(e -> newInstance(clazz, e)).collect(Collectors.toList());
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> long count(AggregatePredicate<AGGREGATE, ?> aggregatePredicate) {
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        return repositorySupervisor.count(predicate);
    }

    @Override
    public <AGGREGATE extends Aggregate<?>> boolean exists(AggregatePredicate<AGGREGATE, ?> aggregatePredicate) {
        Predicate<?> predicate = org.netcorepal.cap4j.ddd.domain.repo.JpaAggregatePredicateSupport.getPredicate((AggregatePredicate<?, ?>)aggregatePredicate);
        return repositorySupervisor.exists(predicate);
    }
}

File: ddd-domain-repo-jpa/src/main/kotlin/com/only4/cap4k/ddd/domain/repo/impl/DefaultAggregateSupervisor.kt
package com.only4.cap4k.ddd.domain.repo.impl

import com.only4.cap4k.ddd.core.domain.aggregate.Aggregate
import com.only4.cap4k.ddd.core.domain.aggregate.AggregateFactorySupervisor
import com.only4.cap4k.ddd.core.domain.aggregate.AggregatePayload
import com.only4.cap4k.ddd.core.domain.aggregate.Id
import com.only4.cap4k.ddd.core.domain.repo.AggregatePredicate
import com.only4.cap4k.ddd.core.domain.repo.AggregateSupervisor
import com.only4.cap4k.ddd.core.domain.repo.RepositorySupervisor
import com.only4.cap4k.ddd.core.share.OrderInfo
import com.only4.cap4k.ddd.core.share.PageData
import com.only4.cap4k.ddd.core.share.PageParam
import com.only4.cap4k.ddd.core.share.misc.resolveGenericTypeClass
import com.only4.cap4k.ddd.domain.repo.JpaAggregatePredicate
import com.only4.cap4k.ddd.domain.repo.JpaAggregatePredicateSupport
import java.util.*

/**
 * 默认聚合管理器
 *
 * @author binking338
 * @date 2025/1/12
 */
class DefaultAggregateSupervisor(
    private val repositorySupervisor: RepositorySupervisor
) : AggregateSupervisor {

    companion object {
        @Suppress("UNCHECKED_CAST")
        private fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> newInstance(
            clazz: Class<AGGREGATE>,
            entity: Any
        ): AGGREGATE =
            try {
                val aggregate = clazz.getDeclaredConstructor().newInstance() as Aggregate<Any>
                aggregate._wrap(entity)
                aggregate as AGGREGATE
            } catch (ex: Exception) {
                throw RuntimeException(ex)
            }

    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY_PAYLOAD : AggregatePayload<ENTITY>, ENTITY : Any> create(
        clazz: Class<AGGREGATE>,
        payload: ENTITY_PAYLOAD
    ): AGGREGATE {
        val entity: ENTITY = AggregateFactorySupervisor.instance.create(payload)
        return newInstance(clazz, entity)
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> getByIds(
        ids: Iterable<Id<AGGREGATE, *>>,
        persist: Boolean
    ): List<AGGREGATE> {
        if (!ids.iterator().hasNext()) {
            return emptyList()
        }

        @Suppress("UNCHECKED_CAST")
        val aggregateClass = resolveGenericTypeClass(
            ids.iterator().next(), 0, Id::class.java, Id.Default::class.java
        ) as Class<AGGREGATE>

        val aggregatePredicate = JpaAggregatePredicate.byIds(
            aggregateClass,
            ids.map { it.value }
        )
        return find(aggregatePredicate, persist)
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> find(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>,
        orders: Collection<OrderInfo>?,
        persist: Boolean
    ): List<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entities = repositorySupervisor.find(pred, orders, persist)
        return entities.map { entity -> newInstance(clazz, entity) }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> find(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>,
        pageParam: PageParam,
        persist: Boolean
    ): List<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entities = repositorySupervisor.find(pred, pageParam, persist)
        return entities.map { entity -> newInstance(clazz, entity) }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> findOne(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>,
        persist: Boolean
    ): Optional<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entity = repositorySupervisor.findOne(pred, persist)
        return if (entity.isPresent) {
            Optional.of(newInstance(clazz, entity.get()))
        } else {
            Optional.empty()
        }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> findFirst(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>,
        orders: Collection<OrderInfo>,
        persist: Boolean
    ): Optional<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entity = repositorySupervisor.findFirst(pred, orders, persist)
        return if (entity.isPresent) {
            Optional.of(newInstance(clazz, entity.get()))
        } else {
            Optional.empty()
        }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> findPage(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>,
        pageParam: PageParam,
        persist: Boolean
    ): PageData<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entities = repositorySupervisor.findPage(pred, pageParam, persist)
        return entities.transform { entity -> newInstance(clazz, entity) }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY: Any> removeByIds(
        ids: Iterable<Id<AGGREGATE, *>>
    ): List<AGGREGATE> {
        if (!ids.iterator().hasNext()) {
            return emptyList()
        }

        @Suppress("UNCHECKED_CAST")
        val aggregateClass = resolveGenericTypeClass(
            ids.iterator().next(), 0, Id::class.java, Id.Default::class.java
        ) as Class<AGGREGATE>

        val aggregatePredicate = JpaAggregatePredicate.byIds(
            aggregateClass,
            ids.map { it.value }
        )
        return remove(aggregatePredicate)
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> remove(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>
    ): List<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entities = repositorySupervisor.remove(pred)
        return entities.map { entity -> newInstance(clazz, entity) }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> remove(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>,
        limit: Int
    ): List<AGGREGATE> {
        val clazz = JpaAggregatePredicateSupport.reflectAggregateClass(predicate)
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        val entities = repositorySupervisor.remove(pred, limit)
        return entities.map { entity -> newInstance(clazz, entity) }
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> count(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>
    ): Long {
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        return repositorySupervisor.count(pred)
    }

    override fun <AGGREGATE : Aggregate<ENTITY>, ENTITY : Any> exists(
        predicate: AggregatePredicate<AGGREGATE, ENTITY>
    ): Boolean {
        val pred = JpaAggregatePredicateSupport.getPredicate(predicate)
        return repositorySupervisor.exists(pred)
    }
}
